### DISCUSSION TITLE
Who's using fast-check? 🤔

### DISCUSSION THREAD

**Discussion by dubzzz** [dubzzz/fast-check#1540] [Show and tell]
As maintainer of fast-check I would be really interested to know who — _individuals, companies or open source projects_ — is using fast-check? For which needs? And eventually the benefits of using it?

**Comment by dan-blank** [comment]
[Individual] I try out fast-check in my free time and want to propose its use in my working place when I feel confident enough to present it. My hope is that it catches bugs in web applications with complex business logic that otherwise would slip through (pretty much the vanilla use case).

**Comment by dubzzz** [comment]
[Company] At Pigment ([website](https://gopigment.com/)/[GitHub](https://github.com/gopigment/)), we use it on all our Pull Requests. It has been introduced quite early in our testing stack and was able to detect problems before we released features to production. It has been useful on various parts of our front (in React): from core algorithms, to invariants on components, but also in things like detection of race conditions. 

Here are some examples of our usages:
- **Rendering of pivoted tables:** In _Pigment_, tables are virtualized for performance reasons. Indeed they may contain thousands and even millions of lines so we don't want to render them all together. As a consequence most of the rendering is handled through div elements absolutely placed. We use property based testing to ensure that whatever the inputs (valid) we pass to our table the resulting DOM will always be a rectangular area with no overlapping between cells (combined with react-testing and jest-dom).
- **Subscribers to back-end services:** Components are offered a way to communicate and watch updates coming from the back. From time to time the updates need an explicit subscription message toward the back but also an unsubscription one. The front deals with pools of registered entities and emit messages toward the back whenever needed: send a subscription whenever the subscriber is the first one for this precise event and an unsubscription whenever it is the last one to disconnect. As a consequence, because everything is asynchronous by nature, we have potential race conditions in mind while working on this part. In order to prevent some to appear we use the scheduler feature of fast-check and check that subscriptions are sent in-time and unsubscriptions too.
- **...many other places:** From algorithms converting back data to front and vice-versa to algorithms transforming structures for optimized look-up and traversal...

Setup:
- We always run them with different seeds in the CI and locally. In other words we never specify the seed.
- We globally define a time limit for property based tests using `fc.configureGlobal({ interruptAfterTimeLimit: 500 })` in order to avoid tests from taking too long. The main motivations behind that one were 1) tests of components take long 2) in case of bug, the shrinker may take some time to shrink so we prefer running the shrinking part locally in order to avoid wasting CI time if the shrinking process is too long.

Did it find bugs?:
Yes! Many times and even when example-based coverage looked well: it helped us to avoid releasing partially broken features.

**Comment by krisanselmo** [comment]
[Company] We use fast-check At Myre ([website](https://www.getmyre.com/)). Especially in our shared engine repository which have abstract/core algorithms and common interfaces/helpers between backends & frontends (all full typescript). To not impact the developer experience, we have separated stages of example based testing (less than 5 sec) and property based testing (~35 sec). Our CI runs both on every pull request on this repo. It helps us to early detect edge cases and bullet proof some complex features. On top of that, we also reuse generated arbitraries to do test fuzzing on our APIs.

**Comment by Vinnl** [comment]
I've introduced it to [@inrupt/solid-client](https://github.com/inrupt/solid-client-js/). The primary use case is for code related to access control, to try to detect situations in which we might cause privilege escalation. It helped us detect and cover a couple of edge cases!

**Comment by jasikpark** [comment]
Using it for a couple of unit tests at [work](https://defined.net), but I haven't attempted to use it for react component tests...

We use storybook and storybook-test for most of our UI-based interaction tests and I haven't sorted out how to integrate fast-check into that setup v.s. just using minifaker to generate data for our tests, and asserting that a form completes or errors with the right values...

It's hard to think how you can do more than unit test with fast-check honestly

**Reply by jasikpark** [reply]
I'll definitely play around with the jest integration as I work through the testing portion of https://epicreact.dev though!

**Comment by gruhn** [comment]
I work for a machine translation company called [DeepL](https://www.deepl.com/translator). The main product is two text boxes where users can translate free text into different languages. There is a lot of complicated stuff going on in these innocent looking text boxes (in the frontend alone), in particular because users can edit both the source text and the translated text, and the fuzzy relationship between the two must constantly be maintained. I use fast-check extensively to hunt down edge cases there. And I think together with the type checker it's really the strongest verification tool (that is still ergonomic) in the entire ecosystem.

**Comment by dubzzz** [comment]
**Companies & major OSS projects verifiably using fast-check**

Every entry was verified against a public asset: a test file on the default branch importing `fast-check` (or `@fast-check/*`) and running real properties, and/or a first-party engineering post. Lockfile-only / transitive dependencies (via `effect`, `syncpack`, `pure-rand`) were excluded. The **Usage** column comes from git-history analysis: scale (property-test files / call sites), sophistication, and whether new property tests were still being added recently.

**Companies**

| Company | Where | Usage | Proof |
|---|---|---|---|
| **Microsoft** | PR-Metrics, retrospectives, AKS tools, physical-ai-toolchain — plus TypeScript parser fuzzing (testimony) & FOSS Fund sponsorship | 🟢 4 repos, ~14 files / 150+ properties — all introduced 2025-11 → 2026-04, still growing | [PR-Metrics](https://github.com/microsoft/PR-Metrics/blob/main/src/task/tests/metrics/codeMetricsData.property.spec.ts) · [AKS fuzz](https://github.com/Azure/vscode-aks-tools/blob/main/src/tests/fuzz/tcpdump-parsing.fuzz.test.ts) · [physical-ai](https://github.com/microsoft/physical-ai-toolchain/blob/main/data-management/viewer/frontend/src/lib/__tests__/playback-utils.property.test.ts) · [FOSS Fund](https://github.com/microsoft/foss-fund/blob/main/README.md) · [TS fuzzing tweet](https://twitter.com/andhaveaniceday/status/1643468822752677888) |
| **Amazon / AWS** | SDK v3, CDK, Smithy, Kiro | 🟢 Advanced & growing — differential fuzzing (`numRuns: 10 000`, `fc.letrec`) added 2026 in SDK/Smithy; CDK properties since **2018**, new files 2024-2026 | [SDK xml fuzz](https://github.com/aws/aws-sdk-js-v3/blob/main/packages-internal/xml-builder/tests/xml-parser.fuzz.spec.ts) · [CDK arbitraries](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/core/test/arbitraries/arbitrary-constructs.ts) · [Smithy CBOR](https://github.com/smithy-lang/smithy-typescript/blob/main/packages/core/src/submodules/cbor/cbor.fast-check.fuzz.spec.ts) · [Kiro blog](https://kiro.dev/blog/property-based-testing-fixed-security-bug/) |
| **MongoDB** | Server repo: vendored fast-check + home-grown PBT framework for the query language (“24 bugs in 8 months”) | 🔥 Heavy · advanced · growing — 65 files, model-based (`fc.commands`) + differential (classic vs SBE engine), 109 commits since 2024-08 | [PBT harness](https://github.com/mongodb/mongo/blob/master/jstests/libs/property_test_helpers/property_testing_utils.js) · [README](https://github.com/mongodb/mongo/blob/master/jstests/libs/property_test_helpers/README.md) |
| **Sentry** | New Sentry CLI/Toolkit — PBT is their primary test strategy; past sponsor (2023-24) | 🔥 Heavy · advanced · growing — **79 files, 689 properties**, model-based SQLite layer (`asyncModelRun`), new files up to 2026-08 | [property test](https://github.com/getsentry/cli/blob/main/packages/cli/test/lib/token-type.property.test.ts) · [model-based](https://github.com/getsentry/cli/blob/main/packages/cli/test/lib/db/model-based.test.ts) |
| **Google** | Magika (AI file-type detection) | ⚪ Light — 1 file / 1 property (2024), maintained | [magika.test.ts](https://github.com/google/magika/blob/main/js/test/magika.test.ts) |
| **OpenAI** | openai-node (partial JSON parser) | ⚪ Light & dormant — 1 file / 2 properties, untouched since 2024-09 | [partial-json test](https://github.com/openai/openai-node/blob/main/tests/_vendor/partial-json-parser/partial-json-parsing.test.ts) |
| **Adobe** | React Spectrum (NumberParser, Color) | 🟡 Light — 2 files since 2021, maintained (last touched 2026-03) | [NumberParser](https://github.com/adobe/react-spectrum/blob/main/packages/%40internationalized/number/test/NumberParser.test.js) · [Color](https://github.com/adobe/react-spectrum/blob/main/packages/react-stately/test/color/Color.test.tsx) |
| **Elastic** | Kibana core (`@fast-check/jest`) | 🟡 Light but fresh — 2 files, introduced 2025-03, extended 2026-06 | [saved-objects test](https://github.com/elastic/kibana/blob/main/src/core/packages/saved-objects/base-server-internal/src/utils/pick_values_based_on_structure.test.ts) |
| **Tiny (TinyMCE)** | Editor monorepo; their Agar test framework wraps fast-check | 🔥 Heavy · advanced · growing — **100 files, 339 properties** since 2021, new files 2026 | [katamari tests](https://github.com/tinymce/tinymce/blob/main/modules/katamari/src/test/ts/atomic/api/arr/ArrBindTest.ts) · [Agar](https://github.com/tinymce/tinymce/blob/main/modules/agar/src/main/ts/ephox/agar/arbitrary/PropertySteps.ts) |
| **Neo4j** | Official JS driver | 🟢 Moderate · advanced · growing — 18 files / 80 properties since 2022, active 2026 | [integer.test.ts](https://github.com/neo4j/neo4j-javascript-driver/blob/master/packages/core/test/integer.test.ts) |
| **DeepL** | CLI (public) + translator frontend (testimony) | 🟢 7 files / 16 properties, brand-new (2026-07) and growing | [tests/property/](https://github.com/DeepLcom/deepl-cli/tree/main/tests/property) · [#1540 testimony](https://github.com/dubzzz/fast-check/discussions/1540) |
| **SAP** | open-ux-tools (2022, dormant) + Luigi (new 2026) | 🟡 Light — 3 files across 2 repos | [store test](https://github.com/SAP/open-ux-tools/blob/main/packages/store/test/unit/secure-store/dummy-store.test.ts) |
| **PostHog** | posthog-js (session IDs, replay triggers) | 🟡 Light — 2 files, since 2021, still touched 2026 | [sessionid.property.test.ts](https://github.com/PostHog/posthog-js/blob/main/packages/browser/src/__tests__/sessionid.property.test.ts) |
| **n8n** | Workflow engine & scheduler | 🟢 7 files / 50 properties, introduced 2026-06, growing | [backoff.property.test.ts](https://github.com/n8n-io/n8n/blob/master/packages/%40n8n/scheduler/src/core/executor/__tests__/backoff.property.test.ts) |
| **Unleash** | OpenAPI schemas & playground | 🟢 13 files / 30 properties since 2022, maintained | [arbitraries.ts](https://github.com/Unleash/unleash/blob/main/src/test/arbitraries.ts) |
| **Rocket.Chat** | Message parser fuzzing | 🟡 1 file, new (2026-04) | [fuzz.test.ts](https://github.com/RocketChat/Rocket.Chat/blob/develop/packages/message-parser/tests/fuzz.test.ts) |
| **Toss** | es-toolkit | 🟡 1 file, new (2026-06) | [chunk.spec.ts](https://github.com/toss/es-toolkit/blob/main/src/array/chunk.spec.ts) |
| **OpenZeppelin** | merkle-tree (`@fast-check/ava`) | 🟢 3 files / 24 properties since 2022, maintained | [core.test.ts](https://github.com/OpenZeppelin/merkle-tree/blob/master/src/core.test.ts) |
| **Lido Finance** | On-chain treasury tooling | 🟢 dedicated `test/fuzz/` suite | [stonks](https://github.com/lidofinance/stonks) |
| **Gusto** | baerly-storage (S3/R2 document DB): GC fuzzing, index crash tests | 🟢 fuzzing-centric suite (`@fast-check/vitest`) | [baerly-storage](https://github.com/Gusto/baerly-storage) |
| **Datawrapper** | Shared numeric/HTML-sanitization lib | 🟢 property tests via `@fast-check/ava` | [datawrapper](https://github.com/datawrapper/datawrapper) |
| **Convex** | Chef (AI app builder): LZ4 implementation | 🟡 1 file | [chef](https://github.com/get-convex/chef) |
| **Enso** | GUI + Yjs collaborative sync layer | 🟢 2 packages | [enso](https://github.com/enso-org/enso) |
| **Input Output (Cardano)** | UTxO coin-selection invariants | 🟢 Advanced but stable — since 2021, maintained | [InputSelection test](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/input-selection/test/InputSelectionPropertyTesting.test.ts) |
| **Agoric** | agoric-sdk + Endo (`@fast-check/ava`) | 🟢 8 files across both, active 2026 | [amountProperties](https://github.com/Agoric/agoric-sdk/blob/master/packages/ERTP/test/unitTests/amountProperties.test.js) · [Endo](https://github.com/endojs/endo/blob/master/packages/marshal/test/rankOrder.test.js) |
| **Prisma** | Client decimal precision (7.x line; ported away in v8) | 🟡 1 file since 2022 | [decimal test.prop](https://github.com/prisma/prisma/blob/7.9.1/packages/client/tests/functional/decimal/precision/tests.ts) |
| **Medplum** | FHIRPath fuzzing | 🟡 1 file, growing (2025→2026) | [fuzz.test.ts](https://github.com/medplum/medplum/blob/main/packages/core/src/fhirpath/fuzz.test.ts) |
| **Inrupt** (Solid) | solid-client-js — access-control safety | 🟡 2 files since 2020, maintained | [datatypes.test.ts](https://github.com/inrupt/solid-client-js/blob/main/src/datatypes.test.ts) |
| **Shiguredo** 時雨堂 | Media-over-QUIC & WebRTC devtools; past sponsor | 🔥 Heavy & growing — 25 files / 172 properties, whole test suites migrated to PBT in 2026 | [moqt-js](https://github.com/shiguredo/moqt-js/blob/develop/src/varint.prop.ts) |
| **Marmelab** | Used on their Limonade project — found real bugs (first-party post, 2019) | 🟡 blog-attested | [blog post](https://marmelab.com/blog/2019/04/18/property-based-testing-js.html) |
| **KAKEHASHI** | Musubi product (first-party post, real bugs found) | 🟡 blog-attested (2024) | [tech blog (JA)](https://kakehashi-dev.hatenablog.com/entry/2024/03/12/110000) |
| **Defined Networking** | Nebula docs tooling; current sponsor | ⚪ 1 file, dormant since 2023 — but sponsoring since then | [validateHostname.test.ts](https://github.com/DefinedNet/nebula-docs/blob/main/docs/guides/using-lighthouse-dns/validateHostname.test.ts) |
| **Consensys (Truffle)** | ABI arbitraries & encoder tests — project sunset | ⚪ Legacy — 8 advanced files, frozen since 2023 | [arbitrary.ts](https://github.com/trufflesuite/truffle/blob/master/packages/abi-utils/lib/arbitrary.ts) |

Also verified but private-code only: **Pigment** (fast-check on every PR — [#1540](https://github.com/dubzzz/fast-check/discussions/1540)), **Criteo** (2018 [blog posts](https://medium.com/criteo-engineering/introduction-to-property-based-testing-f5236229d237)), **Reason Studios**, **Myre**, **MedFlyt**, **LeapYear**. **Datadog** doesn’t test with fast-check but officially *supports* it in Test Optimization (dd-trace-js normalizes `@fast-check/jest` seed suffixes).

**Major open-source projects (community/foundation-owned)**

| Project | Usage | Proof |
|---|---|---|
| **Effect** | 🔥 fast-check is a **runtime dep & public API** (`effect/FastCheck`, `Schema`→`Arbitrary`) + 14 own test files, growing | [FastCheck.ts](https://github.com/Effect-TS/effect/blob/main/packages/effect/src/testing/FastCheck.ts) |
| **noble/scure crypto** (~130M dl/wk) | 🔥 9 files / 75 properties in noble-curves alone, active 2026 | [secp256k1.test.ts](https://github.com/paulmillr/noble-curves/blob/main/test/secp256k1.test.ts) |
| **ArkType** | 🟢 Ships `@ark/fast-check` (types → arbitraries) — structural integration | [ark/fast-check](https://github.com/arktypeio/arktype/tree/main/ark/fast-check) |
| **Jest** (OpenJS) | 🟡 6 files in `expect`/`diff-sequences` since 2019, maintained | [property test](https://github.com/jestjs/jest/blob/main/packages/expect/src/__tests__/matchers-toStrictEqual.property.test.ts) |
| **TanStack DB** | 🟢 6 files / **121 properties**, new (2026), growing | [comparison.property.test.ts](https://github.com/TanStack/db/blob/main/packages/db/tests/comparison.property.test.ts) |
| **Immutable.js** | 🟢 13 files / 32 properties, active again in 2026 | [List.ts](https://github.com/immutable-js/immutable-js/blob/main/__tests__/List.ts) |
| **RxJS** | 🟢 Observable-polyfill lifecycle fuzzing, brand-new (2026-08) | [lifecycle.fuzz.spec.ts](https://github.com/ReactiveX/rxjs/blob/master/packages/observable-polyfill/src/lifecycle.fuzz.spec.ts) |
| **dinero.js** | 🟢 15 files / 75 properties, active | [repo](https://github.com/dinerojs/dinero.js) |
| **undici** (Node.js) | 🟡 Dedicated `test:fuzzing` script (2024), maintained | [fuzzing.test.js](https://github.com/nodejs/undici/blob/main/test/fuzzing/fuzzing.test.js) |
| **remeda** | 🟡 New `*.test-prop.ts` suite + internal PBT guide (2026) | [groupBy.test-prop.ts](https://github.com/remedajs/remeda/blob/main/packages/remeda/src/groupBy.test-prop.ts) |
| **js-yaml** | 🟡 Round-trip fuzzing since 2018, still touched 2026 | [dump-fuzzy.test.mjs](https://github.com/nodeca/js-yaml/blob/master/test/core/dump-fuzzy.test.mjs) |
| **Ramda** | ⚪ 11 files but dormant (est. 2016, last 2024) | [test/equals.js](https://github.com/ramda/ramda/blob/master/test/equals.js) |
| **fp-ts / io-ts** | ⚪ Law-testing pioneers, now dormant (maintenance mode) | [io-ts Arbitrary.ts](https://github.com/gcanti/io-ts/blob/master/test/Arbitrary.ts) |
| **query-string / yaml / Deno std / libauth** | ⚪ Historic single-suite users, dormant | [properties.js](https://github.com/sindresorhus/query-string/blob/main/test/properties.js) · [properties.ts](https://github.com/eemeli/yaml/blob/main/tests/properties.ts) · [compare_with_rust.ts](https://github.com/denoland/std/blob/main/cli/_tools/compare_with_rust.ts) · [libauth](https://github.com/bitauth/libauth) |
| **cron / unhead / polkadot-api / typage** | 🟡 Light but real fuzz/property suites | [cron.fuzz.ts](https://github.com/kelektiv/node-cron/blob/main/tests/cron.fuzz.ts) · [unhead fuzz](https://github.com/unjs/unhead/blob/main/packages/unhead/test/unit/fuzz.test.ts) · [storage.spec.ts](https://github.com/polkadot-api/polkadot-api/blob/main/packages/substrate-bindings/tests/storage.spec.ts) · [typage](https://github.com/FiloSottile/typage/blob/main/tests/property.test.ts) |

**Ecosystem signals**: Vitest’s official docs [recommend fast-check](https://main.vitest.dev/guide/learn/testing-in-practice); [javascript-testing-best-practices](https://github.com/goldbergyoni/javascript-testing-best-practices) (24k⭐) features it; StrykerJS runs a dedicated fast-check e2e integration; active connectors include `zod-fast-check`, `@traversable/zod-test`, `typespec-fast-check` (Stainless), `fp-ts-laws`. Taught at CMU (17-313), used in research tooling at Univ. of Passau (Whisker), featured at TSKaigi 2025.

---

<sub>**Disclaimer** — This analysis was produced by Claude (Anthropic), at my request, over ~80 automated research agents. Scope: full sponsor history (GitHub Sponsors + OpenCollective, via the `dubzzz/sponsors-svg` git history, 2022–2026), repo scans of ~120 corporate GitHub organizations, npm reverse-dependency sweeps ranked by weekly downloads, mining of this repo’s issues/discussions (#1540 among others), and searches across engineering blogs, conference/meetup talks, academic papers and community forums. Every table entry was verified against a fetched public asset — a test file on the default branch actually importing `fast-check` and running properties, or a first-party engineering post; lockfile-only and transitive dependencies (typically via `effect`, `syncpack`, or `pure-rand`’s funding URL) were systematically excluded. Usage grades (scale / sophistication / activity trend) come from git-history analysis of each repo. Errors remain possible — corrections welcome in this thread.</sub>
